packages/compass-preferences-model/src/storage.ts

```diff
@@ -0,0 +1,93 @@
+import { promises as fs } from 'fs';
+import { join } from 'path';
+import type { UserPreferences } from './preferences';
+
+const ENCODING_UTF8 = 'utf8';
+
+export abstract class BasePreferencesStorage {
+  abstract setup(): Promise<void>;
+  abstract getPreferences(): UserPreferences;
+  abstract updatePreferences(
+    attributes: Partial<UserPreferences>
+  ): Promise<void>;
+}
+
+export class SandboxPreferences extends BasePreferencesStorage {
+  constructor(private preferences: UserPreferences) {
+    super();
+  }
+
+  getPreferences() {
+    return this.preferences;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/require-await
+  async updatePreferences(attributes: Partial<UserPreferences>) {
+    this.preferences = {
+      ...this.preferences,
+      ...attributes,
+    };
+  }
+
+  async setup() {
+    // noop
+  }
+}
+
+export class StoragePreferences extends BasePreferencesStorage {
+  private readonly folder = 'AppPreferences';
+  private readonly file = 'General.json';
+  private readonly defaultPreferences: UserPreferences;
+
+  constructor(private preferences: UserPreferences, private basepath?: string) {
+    super();
+    this.defaultPreferences = preferences;
+  }
+
+  private getFolderPath() {
+    return join(this.basepath ?? '', this.folder);
+  }
+
+  private getFilePath() {
+    return join(this.getFolderPath(), this.file);
+  }
+
+  async setup() {
+    // Ensure folder exists
+    await fs.mkdir(this.getFolderPath(), { recursive: true });
+
+    try {
+      this.preferences = JSON.parse(
+        (await fs.readFile(this.getFilePath(), ENCODING_UTF8)).toString()
```

Comment:
```suggestion
        (await fs.readFile(this.getFilePath(), ENCODING_UTF8))
```

or

```suggestion
        (await fs.readFile(this.getFilePath(), 'utf8'))
```

(Typescript should be aware of the different encodings and know that calling `readFile` with `'utf8'` will return a string)